core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -1350,6 +1458,24 @@ private Class<?> getContainerType(Type parameterizedType) {
             return null;
         }
 
+        @Nullable
+        private Class<?> getRawContainerType(Type parameterizedType) {
+            final Class<?> rawType = toRawType(parameterizedType);
+            if (rawType == Iterable.class ||
+                rawType == List.class ||
+                rawType == Collection.class ||
+                rawType == Set.class ||
+                rawType == Map.class ||
+                Iterable.class.isAssignableFrom(rawType) ||
+                List.class.isAssignableFrom(rawType) ||
+                Collection.class.isAssignableFrom(rawType) ||
+                Set.class.isAssignableFrom(rawType) ||
+                Map.class.isAssignableFrom(rawType)) {
+                return rawType;
+            }
+            return null;
```

Comment:
Question) I didn't understand the intention behind this method. Couldn't we just return the raw type as is?

e.g.
 ```suggestion
            return toRawType(parameterizedType);
```

Reply:
Good Point! 

This method is to get real container type. because `getContainerType()` return `ArrayList.class`. However, If input value has `ImmutableList.class`, `ImmutableList` cannot be cast to `ArrayList.class`.  Thus we need to get `rawContainerType` instead of `containerType`. 

With your suggestion, this function will return `int.class` if input type is `int.class`, and it is not `container type` actually. (it's `elementType`!). 


`rawContainerType` is used to validate whether if it can be cast to in here!
```java
    private static BiFunction<AnnotatedValueResolver, ResolverContext, Object>
    attributeResolver(Iterable<AttributeKey<?>> attrKeys) {
        return (resolver, ctx) -> {
            StringBuilder errorMsgBuilder = null;
            Class<?> targetType = resolver.rawContainerType();
            if (targetType == null) {
                targetType = resolver.elementType().isPrimitive() ?
                              Primitives.wrap(resolver.elementType())
                              : resolver.elementType();
            }

            for (AttributeKey<?> attrKey : attrKeys) {
                final Object value = ctx.context.attr(attrKey);
                if (value != null) {
                    final boolean isValidType = targetType.isInstance(value);

                    if (isValidType) {
                        return value;

```

https://github.com/chickenchickenlove/armeria/blob/5874745496084408008aaf0a73925a7a5c711965/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java#L866-L883

Reply:
I see, I was rather thinking that `rawContainerType` is a very specific function that I don't think other methods will use. Rather we could define a `rawType` instead.

e.g.
https://github.com/jrhee17/armeria/commit/81e683642430d2fec0f7ceeeea1b84dec2849ee0

Reply:
@jrhee17 nim, thanks for your comments. 
super good point! 🚀 🚀 
I applied it!